internal/fuse/snapshots_dir.go

```diff
@@ -15,186 +17,183 @@ import (
 	"bazil.org/fuse/fs"
 )
 
-// SnapshotsDir is a fuse directory which contains snapshots named by timestamp.
-type SnapshotsDir struct {
-	inode   uint64
-	root    *Root
-	names   map[string]*restic.Snapshot
-	latest  string
-	tag     string
-	host    string
-	snCount int
-
-	template string
+// SnapshotsDirStructure contains the directory structure for snapshots
+// it uses a paths and time template to generate a map of pathnames
+// pointing to the actual snapshots. For templates that end with a time,
+// also "latest" links are generated.
+type SnapshotsDirStructure struct {
+	root          *Root
+	pathTemplates []string
+	timeTemplate  string
+
+	names      map[string]*restic.Snapshot
+	latest     map[string]string
+	latestTime map[string]time.Time
+	snCount    int
+	lastCheck  time.Time
 }
 
-// SnapshotsIDSDir is a fuse directory which contains snapshots named by ids.
-type SnapshotsIDSDir struct {
-	inode   uint64
-	root    *Root
-	names   map[string]*restic.Snapshot
-	snCount int
+// NewSnapshotsDir returns a new directory containing snapshots.
+func NewSnapshotsDirStructure(root *Root, pathTemplates []string, timeTemplate string) *SnapshotsDirStructure {
+	return &SnapshotsDirStructure{
+		root:          root,
+		pathTemplates: pathTemplates,
+		timeTemplate:  timeTemplate,
+		snCount:       -1,
+	}
 }
 
-// HostsDir is a fuse directory which contains hosts.
-type HostsDir struct {
-	inode   uint64
-	root    *Root
-	hosts   map[string]bool
-	snCount int
+func (d *SnapshotsDirStructure) uniqueName(name string) (newname string) {
+	newname = name
+	for i := 1; ; i++ {
+		if _, ok := d.names[newname]; !ok {
+			break
+		}
+		newname = fmt.Sprintf("%s-%d", name, i)
+	}
+	return newname
 }
 
-// TagsDir is a fuse directory which contains tags.
-type TagsDir struct {
-	inode   uint64
-	root    *Root
-	tags    map[string]bool
-	snCount int
+// replaceMultiple replaces pattern in all paths strings by repl strings.
+// If pattern does not exist in a string, the string remains unchanged.
+// If pattern exists, for each repl string a new path with replaced patterns will be added
+// Note that if the pattern exists in a string and repl is empty, the string will be removed.
+func replaceMultiple(paths []string, pattern string, repl []string) (replacedPaths []string) {
```

Comment:
I find the logic that uses this function hard to follow. Why not have a function that formats a single filename, by just looping over the content?

```go
func formatPath(template string, sn *restic.Snapshot) string {
	var (
		inVerb = false
		out    strings.Builder
	)
	for _, c := range template {
		if !inVerb {
			if c == '%' {
				inVerb = true
			} else {
				out.WriteRune(c)
			}
			continue
		}

		switch c {
		// handle the various formatting verbs
		}
		inVerb = false
	}

	return out.String()
}
```

Reply:
Thanks for the suggestion! It is however a bit more complicated: There can be multiple tags resulting in more than one path and the "latest" link should be handled. I hence extended your code example a bit...
Hope it is now much clearer!